x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -357,18 +367,101 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
                 bulkRequest,
                 TransportSingleItemBulkWriteAction.<IndexResponse>wrapBulkResponse(ActionListener.wrap(indexResponse -> {
                     assert docId.equals(indexResponse.getId());
-                    // TODO: replace with actual domain information
                     final VersionedDocument versionedDocument = new VersionedDocument(
                         profileDocument,
                         indexResponse.getPrimaryTerm(),
                         indexResponse.getSeqNo()
                     );
                     listener.onResponse(versionedDocument.toProfile(Set.of()));
-                }, listener::onFailure))
+                }, e -> {
+                    if (e instanceof VersionConflictEngineException) {
+                        // Document already exists with the specified ID, get the document with the ID
+                        // and check whether it is the right profile for the subject
+                        getOrCreateProfileWithBackoff(subject, profileDocument, DEFAULT_BACKOFF.iterator(), listener);
+                    } else {
+                        listener.onFailure(e);
+                    }
+                }))
             )
         );
     }
 
+    // Package private for test
+    void getOrCreateProfileWithBackoff(
+        Subject subject,
+        ProfileDocument profileDocument,
+        Iterator<TimeValue> backoff,
+        ActionListener<Profile> listener
+    ) {
+        getVersionedDocument(profileDocument.uid(), ActionListener.wrap(versionedDocument -> {
+            if (versionedDocument == null) {
+                // Document not found. This can happen if the GET request hits a replica that is still processing the document
+                if (backoff.hasNext()) {
+                    final TimeValue backoffTimeValue = backoff.next();
+                    logger.debug("retrying get profile document [{}] after [{}] backoff", profileDocument.uid(), backoffTimeValue);
+                    client.threadPool()
+                        .schedule(
+                            () -> getOrCreateProfileWithBackoff(subject, profileDocument, backoff, listener),
+                            backoffTimeValue,
+                            ThreadPool.Names.GENERIC
+                        );
+                } else {
+                    // Retry has depleted. This can only happen when the document or the profile index itself gets deleted
+                    // in between requests.
+                    listener.onFailure(
+                        new ElasticsearchException("failed to retrieving profile [{}] after all retries", profileDocument.uid())
+                    );
+                }
+                return;
+            }
```

Comment:
Do you think we could split this `getVersionedDocument` method in two parts here, and move the second part below this comment in the first's listener, such that the first part has the sole responsability of gettting the profile doc with retry.

Reply:
I tried to split them but at the end decided to keep it as is. The main reason is `getOrCreateProfileWithBackoff` is called in two places and it expects the same logic (the 2nd part) for handling an existing document. If we split the 2nd part out, we have to ensure the listener argument is wrapped properly to always ensure the 2nd part is included. It is doable. But I think it does not add too much value since the method isn't too long at this stage. And it has the future risk to change one call site without changing the other one (I actually made this mistake when refactoring it).